### DISCUSSION TITLE
Is it possible dropping `--debugger` and `--debugger-pipe` when debugging cmake scripts?

### DISCUSSION THREAD

**Discussion by zchrissirhcz** [microsoft/vscode-cmake-tools#3891] [Q&A] [answered]
I use CMake Tools extension to debug a cmake script file, `test.cmake` by writing a `launch.json` file.

When debugging, `test.cmake` receives 3 extra arguments:
```
--debugger
--debugger-pipe
\\.\\pipe\\cmake-debugger-pipe\\9568e66a-5b26-4573-b376-551ad2c48672
```

In fact, my actual `test.cmake` only accept some arguments, e.g. `update`, `install`, `search`, etc, neither `--debugger` nor `--debugger-pipe` is valid arguments. Can I drop the extra arguments?

I read the document https://github.com/microsoft/vscode-cmake-tools/blob/main/docs/debug.md , but not found solution.

My `.vscode/launch.json` is:
```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "type": "cmake",
            "request": "launch",
            "name": "CMake: Configure",
            "cmakeDebugType": "script",
            "scriptPath": "${workspaceFolder}/test.cmake",
            "scriptArgs": [
                "-DHELLO=WORLD",
                "-DCMAKE_BUILD_TYPE=Debug"
            ],
            "clean": false,
            "configureAll": false
        }
    ]
}
```

A sample `test.cmake` is:
```cmake
message("Welcom to CMake script mode!")
message("Hello, ${HELLO}")

message(STATUS "CMAKE_ARGC: ${CMAKE_ARGC}")
foreach(_arg RANGE ${CMAKE_ARGC})
  message(STATUS "${_arg}: ${CMAKE_ARGV${_arg}}")
endforeach()

function(add_numbers num1 num2 OUTPUT_VAR)
  math(EXPR output "${num1} + ${num2}")
  set(${OUTPUT_VAR} ${output} PARENT_SCOPE)
endfunction()

set(my_result 0)

add_numbers(5 3 my_result)

message("5 + 3 = ${my_result}")
message("bye~")
```

The output of debugging `test.cmake` is:
![image](https://github.com/microsoft/vscode-cmake-tools/assets/3831847/67d6c5f7-f433-4edd-872b-557dd6d360e7)

**Comment by zchrissirhcz** [comment]
Just found the source where extra args are pushed:

<del> https://github.com/microsoft/vscode-cmake-tools/pull/3181/files#diff-debb2b28d6617ee810ad4ca1171ca3ba817ec0aa42dfe921fbfc541c53444412R242-R245 </del>

https://github.com/microsoft/vscode-cmake-tools/blob/main/src/debug/debuggerScriptDriver.ts#L25-L30

![image](https://github.com/microsoft/vscode-cmake-tools/assets/3831847/7731e354-0f95-435e-b0f2-156125070b60)



@gcampbell-msft Is it possible that I remove extra args when debug CMake script?

![image](https://github.com/microsoft/vscode-cmake-tools/assets/3831847/8d34b864-c19f-498c-b4cc-6030f5612b92)

**Comment by gcampbell-msft** [comment, marked as answer]
@zchrissirhcz No, the way that the cmake debugger works is that you have to pass these arguments to be able to debug. Is there some other scenario that you're trying to complete that we don't support?

**Reply by zchrissirhcz** [reply]
@gcampbell-msft 
In my scenario, a startup cmake script is invoded in script mode, which will parse command line arguments and form a new commands that invoke regular cmake configure. The startup cmake script, let's call it `mybuild.cmake`, only accept some command line arguments, and `--debugger` is considered invalid, then it exit. I would like to drop `--debugger` and `--debugger-pipe`, etc, only passing supported arguments to `mybuild.cmake`.

**Reply by gcampbell-msft** [reply]
How are you expecting to debug the script then? If we don't include `--debugger` or `--debugger-pipe`, then we will not be able to debug. 

In order to debug CMake, whether through Configure or Script mode, it requires those arguments to be able to debug. 

Does that make sense?

**Reply by zchrissirhcz** [reply]
OK, `--debugger` or `--debugger-pipe` is for debugging, they are required. In order to debug CMake, for script mode, is it possible that in the debugee (`mybuild.cmake`), the `CMAKE_ARGV${_arg}` does not includes `--debugger` and `--debugger-pipe`?

**Reply by zchrissirhcz** [reply]
It seem like `concreteArgs` is received in `mybuild.cmake`, but I expect it only receives `scriptArgs`.

**Reply by zchrissirhcz** [reply]
It seems I was wrong. I should read the CMake document, the `--debugger` and `--debugger-pipe` is required by CMake since 3.27.

https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-debugger

**Reply by gcampbell-msft** [reply]
> OK, `--debugger` or `--debugger-pipe` is for debugging, they are required. In order to debug CMake, for script mode, is it possible that in the debugee (`mybuild.cmake`), the `CMAKE_ARGV${_arg}` does not includes `--debugger` and `--debugger-pipe`?

For this, I don't think that this can be done, because we are assuming the `mybuild.cmake` is what is desired to be debugged, and in that case the debugger arguments are required. To my knowledge, there is no way to pass arguments without them appearing in the CMAKE_ARGV.

**Reply by zchrissirhcz** [reply]
I think dropping `--debugger` and `--debugger-pipe` is possible. This should be done inside cmake, instead of VSCode CMake Tools extension.

Let me give a similar usage. When using GDB for C/C++ code debugging, and the `main.cpp` requires an argument, say, an integer.

The command to start GDB with arguments is:
```bash
gdb --args a.out 42  
```

And in `main.cpp`, it receives `a.out` and `42`, but it won't receive `--args`.

```cpp
#include <iostream>  
#include <cstdlib>  // For std::atoi  

int main(int argc, char* argv[]) {  
    // Ensure that we received one argument  
    if (argc != 2) {  
        std::cerr << "Usage: " << argv[0] << " <integer_value>" << std::endl;  
        return 1;  
    }  

    // Convert the received argument to an integer  
    int value = std::atoi(argv[1]);  

    // Print the integer value  
    std::cout << "The provided integer value is: " << value << std::endl;  

    return 0;  
}
```

**Reply by gcampbell-msft** [reply]
@zchrissirhcz I can confirm that this would need to be an improvement in cmake itself, because right now even if we put the --debugger or --debugger-pipe argument in front of `-P <script>`, the script still receives the arguments. 

I would direct you to the Kitware site: https://gitlab.kitware.com/cmake/cmake/-/issues, to create an issue there. Thanks!
